src/uct/ib/dc/dc_mlx5_ep.h

```diff
@@ -567,6 +570,69 @@ uct_dc_mlx5_iface_dci_detach(uct_dc_mlx5_iface_t *iface, uct_dc_mlx5_ep_t *ep)
     return 1;
 }
 
+static inline ucs_status_t
+uct_dc_mlx5_dci_pool_add_dci(uct_dc_mlx5_iface_t *iface, uint8_t pool_index,
+                             uint8_t dci_index)
+{
+    uct_dc_mlx5_dci_pool_t *pool = &iface->tx.dci_pool[pool_index];
+    uct_dc_mlx5_dci_config_t pool_config = {
+        .u64 = pool->config_key
+    };
+    ucs_status_t status;
+
+
+    if ((pool->size == pool->capacity) ||
+        iface->tx.dcis[dci_index].initialized) {
+        return UCS_OK;
+    }
+
+    status = uct_dc_mlx5_iface_create_dci(
+            iface, pool_index, dci_index,
+            pool_config.key.path_index,
+            iface->flags & UCT_DC_MLX5_IFACE_FLAG_DCI_FULL_HANDSHAKE);
+
+    if (status != UCS_OK) {
+        ucs_error("iface %p: failed to create dci %u at pool %u", iface,
+                  dci_index, pool_index);
+        return status;
+    }
+
+    pool->stack[pool->size] = dci_index;
+    ++pool->size;
+    return UCS_OK;
+}
+
+static inline ucs_status_t uct_dc_mlx5_iface_dci_alloc(uct_dc_mlx5_iface_t *iface, uct_dc_mlx5_ep_t *ep)
+{
+    /* take a first available dci from stack.
+     * There is no need to check txqp because
+     * dci must have resources to transmit.
+     */
+    uint8_t pool_index           = uct_dc_mlx5_ep_pool_index(ep);
+    uct_dc_mlx5_dci_pool_t *pool = &iface->tx.dci_pool[pool_index];
+    ucs_status_t status;
+
+    status = uct_dc_mlx5_dci_pool_add_dci(iface, pool_index, iface->tx.dci_counter);
```

Comment:
instead of adding this on fast path,check pool size in uct_dc_mlx5_iface_dci_can_alloc and in needed extend the pool there